contracts/core.clar

```diff
@@ -0,0 +1,180 @@
+(define-constant ERR_CONTRACT_DO_NOT_EXISTS u1002)
+(define-constant ERR_VOTE_HAS_ENDED u1003)
+(define-constant ERR_VOTE_STILL_IN_PROGRESS u1004)
+
+;; TODO: think about replacing with buff
+(define-constant STATE_DEFINED u0)
+(define-constant STATE_STARTED u1)
+(define-constant STATE_LOCKED_IN u2)
+(define-constant STATE_ACTIVE u3)
+(define-constant STATE_FAILED u4)
+
+(define-constant DEFAULT_VOTING_PERIOD u200)
+(define-constant DEFAULT_VOTING_THRESHOLD u90)
+
+(define-data-var cityWallet principal 'ST31270FK25JBGCRWYT7RNK90E946R8VW6SZYSQR6)
+
+(define-data-var activeMiningContract principal .citycoin)
+
+;; used as MiningContract ID and key in Voting map.
+(define-data-var miningContractNonce uint u0)
+
+(define-map MiningContracts
+  principal  ;; MiningContract address
+  { id: uint, state: uint }
+)
+
+(define-map MiningContractVotes
+  uint ;; MiningContract id
+  {
+    address: principal,
+    startBH: uint,
+    endBH: uint,
+    miners: uint,
+    votes: uint
+  }
+)
+
+(define-map MiningCandidateVoters
+  { contract: principal, voter: principal }
+  bool
+)
+
+(define-read-only (has-voted-on-candidate (contract principal))
+  (is-some (map-get? MiningCandidateVoters
+    { contract: contract, voter: contract-caller }
+  ))
+)
+
+(define-read-only (get-city-wallet)
+  (var-get cityWallet)
+)
+
+(define-public (set-city-wallet (newCityWallet principal))
+  (begin
+    (asserts! (is-authorized) (err ERR_UNAUTHORIZED))
+    (ok (var-set cityWallet newCityWallet))
+  )
+)
+
+(define-read-only (get-mining-contract (address principal))
+  (map-get? MiningContracts address)
+)
+
+(define-public (add-mining-contract (address principal))
+  (let
+    (
+      (newNonce (+ (var-get miningContractNonce) u1))
+    )
+    (asserts! (is-authorized) (err ERR_UNAUTHORIZED))
+    (asserts! (not (is-eq (get-active-mining-contract) address)) (err ERR_CONTRACT_ALREADY_EXISTS))
+    (asserts! (is-none (get-mining-contract address)) (err ERR_CONTRACT_ALREADY_EXISTS))
+    
+    (var-set miningContractNonce newNonce)
+    (map-set MiningContractVotes 
+      newNonce
+      {
+        address: address,
+        startBH: (+ block-height u1),
+        endBH: (+ block-height u1 DEFAULT_VOTING_PERIOD),
+        miners: u0,
+        votes: u0
+      }
+    )
+    (ok (map-set MiningContracts address { id: newNonce, state: STATE_DEFINED } ))
+  )
+)
+
+(define-read-only (get-mining-contract-vote (contractId uint))
+  (map-get? MiningContractVotes contractId)
+)
+
+(define-public (vote-on-mining-contract (contract principal))
+  (let
+    (
+      (contractId (get id (unwrap! (get-mining-contract contract) (err ERR_CONTRACT_DO_NOT_EXISTS))))
+      (contractVote (unwrap-panic (get-mining-contract-vote contractId)))
+    )
+    (asserts! (is-between block-height (get startBH contractVote) (get endBH contractVote)) 
+      (err ERR_VOTE_HAS_ENDED))
+    (asserts! (not (has-voted-on-candidate contract))
+      (ok false))
```

Comment:
Why return `ok false` here over throwing an error like `ERR_ALREADY_VOTED` or similar?

Reply:
Great finding.
At first I wanted to suppress all errors in this function as they would fail `mine-tokens` that will be responsible for calling it, but later I changed my mind and forgot to fix this execution path.
I'll add proper error with next commit.

Reply:
Ah... that's interesting! I hadn't thought of it through the path of calling via `mine-tokens` and instead was looking at it as an independent function.

I think this still makes sense to error out though since it follows the same pattern as all of our other checks and balances. The UI can be updated to check `has-voted-on-candidate` before displaying a vote button or similar.

Reply:
Fiexed with https://github.com/citycoins/citycoin/pull/111/commits/0222f6310e5c9745dbc4360482ae21294be04035